src/coreclr/jit/lclmorph.cpp

```diff
@@ -1353,6 +1353,77 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
                 break;
             }
 
+            case GT_INDEX_ADDR:
+            {
+                assert(TopValue(2).Node() == node);
+                assert(TopValue(1).Node() == node->gtGetOp1());
+                assert(TopValue(0).Node() == node->gtGetOp2());
+
+                // We only expect TopValue(1).IsAddress() to be true for stack allocated arrays
+                if (node->gtGetOp2()->IsCnsIntOrI() && TopValue(1).IsAddress())
```

Comment:
I think this is too conservative. `GT_INDEX_ADDR` should never expose its op1 as it implies a bounds check.

Take a look at `IDEAEncryption:de_key_idea`: the array ends up exposed which blocks propagation of the length, which leads to regressions.

Reply:
I'm planning to move this to the global morph.

Reply:
How is that going to help? If the array gets marked as exposed during local morph then the JIT will be forced to make very pessimistic assumptions about the contents.

We need to get the same sort of guarantees for a stack array that we have for a heap array. The codegen touching the array should be similar or better (we may see some extra prolog code to zero the array, so some size regressions are expected). So we need to dig into cases where it's not that way yet.

If all array indices are known we might expect to see promotion of array elements. So it would be good to verify that this works in simple cases.

If all array indices are known good things happen, but if any index is not known then things do not go as well. For example:

```C#
    public static int f()
    {
        int[] a = [1, 2];
        int i = 0;
        int j = 1;

        return a[i] + a[j];
    }
```
Currently this gives:
```asm
; Method f():int (FullOpts)
G_M32463_IG01:  ;; offset=0x0000
       sub      rsp, 56
       vxorps   xmm4, xmm4, xmm4
       vmovdqa  xmmword ptr [rsp+0x20], xmm4
       xor      eax, eax
       mov      qword ptr [rsp+0x30], rax
						;; size=21 bbWeight=1 PerfScore 3.83

G_M32463_IG02:  ;; offset=0x0015
       mov      rax, 0x7FF7E0BEECC8      ; int[]
       mov      qword ptr [rsp+0x20], rax
       mov      dword ptr [rsp+0x28], 2
       mov      dword ptr [rsp+0x30], 1
       mov      dword ptr [rsp+0x34], 2
       lea      rax, [rsp+0x20]
       mov      eax, dword ptr [rax+0x08]
       test     eax, eax
       je       SHORT G_M32463_IG04
       mov      ecx, dword ptr [rsp+0x30]
       cmp      eax, 1
       jbe      SHORT G_M32463_IG04
       mov      eax, ecx
       add      eax, dword ptr [rsp+0x34]
						;; size=66 bbWeight=1 PerfScore 12.50

G_M32463_IG03:  ;; offset=0x0057
       add      rsp, 56
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25

G_M32463_IG04:  ;; offset=0x005C
       call     CORINFO_HELP_RNGCHKFAIL
       int3     
						;; size=6 bbWeight=0 PerfScore 0.00
; Total bytes of code: 98
```

Reply:
Seems like local morph doesn't really have a concept for an indexable local. I think we can get past this without adding a bunch of new concepts, but we'll see.

Reply:
Not going to be quite as easy as I'd hoped. The code you have handles the "known index" by flattening it to a GT_LCL_FLD. But if the index is not known, it seems like we have two choices:
1. Keep the current INDEX_ADDR(GT_LCL_VAR_ADDR, index) and teach lcl morph, promotion, etc that there is now anther case besides ret buffs where GT_LCL_ADDR does not imply exposure; or
2. Introduce new concepts like GT_LCL_VAR_INDEX and STORE_GT_LCL_VAR_INDEX to encapsulate the entirety of the access and figure out how long those need to exist and how well they mesh with our other array access operations and such. Seems like not long after this we will want GT_LCL_VAR_INDEX_FLD to handle arrays of structs.

(1) is less disruptive but adds annoying context to GT_LCL_VAR_ADDR. We can annotate the nodes that happen under INDEX_ADDR and also mark the local in some way. That's the path I started down, so perhaps it's worth hacking it through to the "end" just to see how much else is impacted.

Reply:
One snag with (1) is not only that the LCL_VAR_ADDR is contextual, but also that the necessary context can be two hops away, eg we need to differentiate IND(INDEX_ADDR(LCL_VAR_ADDR, ...)) from STOREIND(INDEX_ADDR(LCL_VAR_ADDR, ...)).

Reply:
It seem unlikely to me we'll be able to get very far trying to handle unknown indices -- no passes will be able to reason about these loads or stores, so it seems we will effectively have to be as conservative as if we had exposed it.

Is the goal to allow at least reasoning about the length field even when the elements accesses are unknown? I wonder if we could somehow split the array into an exposed and an unexposed part during stack allocation.

Reply:
The goal is to not end up worse off if we stack allocate an array... if we don't stack allocate the above code is:
```asm
; Method A:f():int (FullOpts)
G_M32463_IG01:  ;; offset=0x0000
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25

G_M32463_IG02:  ;; offset=0x0004
       mov      rcx, 0x7FF7E6E6ECC8      ; int[]
       mov      edx, 2
       call     CORINFO_HELP_NEWARR_1_VC
       mov      rcx, 0x200000001
       mov      qword ptr [rax+0x10], rcx
       mov      ecx, dword ptr [rax+0x10]
       add      ecx, dword ptr [rax+0x14]
       mov      eax, ecx
						;; size=42 bbWeight=1 PerfScore 8.00

G_M32463_IG03:  ;; offset=0x002E
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code: 51
```
So a bit more efficient initialization and no bounds checks, and if we knew the array was thread private, perhaps the whole thing would optimize away.

Splitting might be an option, or (perhaps similar) we could model the on-stack array as span.

Reply:
With
```cs
var arr = new int[5];
arr[0] = 1; arr[1] = 2; arr[2] = 3; arr[3] = 4; arr[4] = 5;
Console.WriteLine(arr.Length);
Console.WriteLine(arr[0] + arr[1]);
```

We get 
```asm
G_M27646_IG01:  ;; offset=0x0000
       sub      rsp, 40
                                                ;; size=4 bbWeight=1 PerfScore 0.25
G_M27646_IG02:  ;; offset=0x0004
       mov      ecx, 5
       call     [System.Console:WriteLine(int)]
       mov      ecx, 3
       call     [System.Console:WriteLine(int)]
       nop
                                                ;; size=23 bbWeight=1 PerfScore 6.75
G_M27646_IG03:  ;; offset=0x001B
       add      rsp, 40
       ret
                                                ;; size=5 bbWeight=1 PerfScore 1.25
```

but with
```cs
var arr = new [] { 1, 2, 3, 4, 5 };
Console.WriteLine(arr.Length);
Console.WriteLine(arr[0] + arr[1]);
```

we get
```asm
G_M27646_IG01:  ;; offset=0x0000
       push     rsi
       push     rbx
       sub      rsp, 40
                                                ;; size=6 bbWeight=1 PerfScore 2.25
G_M27646_IG02:  ;; offset=0x0006
       mov      rcx, 0x1D38EBA2988      ; const ptr
       mov      ebx, dword ptr [rcx]
       mov      rcx, 0x1D38EBA298C
       mov      esi, dword ptr [rcx]
       mov      ecx, 5
       call     [System.Console:WriteLine(int)]
       lea      ecx, [rbx+rsi]
       call     [System.Console:WriteLine(int)]
       nop
                                                ;; size=45 bbWeight=1 PerfScore 11.50
G_M27646_IG03:  ;; offset=0x0033
       add      rsp, 40
       pop      rbx
       pop      rsi
       ret
                                                ;; size=7 bbWeight=1 PerfScore 2.25
```

Reply:
I think a good goal here is to minimize the possible downside and not yet worry about maximizing the possible upside. So the comparison should be against the heap allocated code.

> this trivial case can be resolved by handling `INDEX_ADDR` and `ARR_LENGTH` in global morph so that the assertion can be propagated.

It does not seem trivial to me. If the array is address exposed how can morph make any assertions about its contents? Maybe VN can reason about some of this, since it tries to model memory updates, but then it is easily foiled by writes or calls. I can see (perhaps) special casing the length or method table fields, since we know they can't be safely modified.

Reply:
With my latest commit now the codegen of your snippet becomes:
```asm
; Program:f():int (FullOpts)

G_M52058_IG01:  ;; offset=0x0000
       sub      rsp, 24
       vxorps   xmm4, xmm4, xmm4
       vmovdqa  xmmword ptr [rsp], xmm4
       xor      eax, eax
       mov      qword ptr [rsp+0x10], rax
                                                ;; size=20 bbWeight=1 PerfScore 3.83
G_M52058_IG02:  ;; offset=0x0014
       mov      rax, 0x7FFD0E66ECC8      ; int[]
       mov      qword ptr [rsp], rax
       mov      dword ptr [rsp+0x08], 2
       mov      dword ptr [rsp+0x10], 1
       mov      dword ptr [rsp+0x14], 2
       mov      eax, dword ptr [rsp+0x10]
       add      eax, dword ptr [rsp+0x14]
                                                ;; size=46 bbWeight=1 PerfScore 7.25
G_M52058_IG03:  ;; offset=0x0042
       add      rsp, 24
       ret
                                                ;; size=5 bbWeight=1 PerfScore 1.25
; Total bytes of code 71
```

which is better than before, and it's also better than heap allocated array.